Removing the T4 templates used by Vector<T>

tannergooding opened the pull request:
This removes the T4 templates used by `Vector<T>`. The T4 templates did not reliably work with SDK based projects nor shared projects.

This also simplifies the code used by `Vector<T>` to simply use a generic `for` loop and `Unsafe.As`. On platforms with hardware acceleration, they currently support basic loop unrolling when the upper end of the loop is `Vector<T>.Count` (this is enabled via `GTF_ICON_SIMD_COUNT`: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/optimizer.cpp) and the overall codegen shouldn't change for the reflection callback (or for methods where a particular type is not intrinsic).

On platforms where the loop unrolling doesn't exist, the worst case is an ~4x regression. This could likely be mitigated by minimally recognizing `Vector<T>.Count` as intrinsic and enabling the unrolling support; if that is a concern. I provided more numbers here: https://github.com/dotnet/runtime/issues/36160#issuecomment-689868849

This should namely make it simpler to support additional types, such as `nint` and `nuint` (https://github.com/dotnet/runtime/issues/36160).

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
This file was `Vector.cs`

tannergooding replied:
I removed the regions and simplified the code, but otherwise left the general ordering of the file the same as it already was.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:
```diff
@@ -2,4074 +2,938 @@
 // The .NET Foundation licenses this file to you under the MIT license.
```
This file was `Vector_Operations.cs` + some trailing bits from `Vector.cs`, it now just contains `static class Vector`.

tannergooding replied:
This file is largely untouched as it, minus the Widen/Narrow and Same-Size Conversion functions were already not templated. Just the `#regions` were removed.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+    [Intrinsic]
+    public struct Vector<T> : IEquatable<Vector<T>>, IFormattable where T : struct
+    {
+        private ulong _00;
```
The type contains 2x `ulong` now, rather than the explicit layout `Register` struct (which was a 16-byte, 66 field explicit layout struct).

tannergooding replied:
I had to add back `Register`. ARM64 was failing due to some disconnect between the HFA classification causing it to be `hasMultiregStructArgs` and `fgMorphMultiregStructArgs` not reporting the type as `TYP_STRUCT`.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+
+            for (nint index = 0; index < Count; index++)
+            {
+                SetElement(index, value);
```
The logic which uses `Unsafe.Add(Unsafe.As())` is relegated to `GetElement` and `SetElement`.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+        {
+            ThrowHelper.ThrowForUnsupportedVectorBaseType<T>();
+
+            if (values is null)
```
Just a note: We currently enforce this to `throw` if `values` is null, which prevents us from being able to have it just forward to the `ROSpan` version.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+        {
+            Vector<T> result = default;
+
+            for (nint index = 0; index < Count; index++)
```
As with the `IsHardwareAccelerated` path in the previous code, this just does a `for` loop from `index to Count` and calls the relevant `Scalar*` operation.
It just uses `SetElement` and `GetElement` now, rather than using stackalloc and the indexer.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+        /// <param name="factor">The scalar value.</param>
+        /// <returns>The scaled vector.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector<T> operator *(Vector<T> value, T factor)
```
This one and the `T * Vector<T>` variant just below were doing `value * new Vector<T>(factor)`. I've changed that to be a `for` loop instead for the software fallback case.

jkotas replied:
I am not sure whether I would bother with one-offs perf tweaks for the non-accelerated case like this at all. It won't move the needle. I would keep the non-accelerated case as simple as possible everywhere.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+        /// <summary>Negates a given vector.</summary>
+        /// <param name="value">The source vector.</param>
+        /// <returns>The negated vector.</returns>
+        public static Vector<T> operator -(Vector<T> value) => Zero - value;
```
This one and `~` could likely benefit from the same optimization I did for `operator *(Vector<T>, T)` where it does a different `for` loop if not HWAccelerated.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+        public static explicit operator Vector<byte>(Vector<T> value)
+        {
+            ThrowHelper.ThrowForUnsupportedVectorBaseType<T>();
+            return Unsafe.As<Vector<T>, Vector<byte>>(ref value);
```
This was previously doing `return new Vector<byte>(ref value.register);` and I've changed it to just `Unsafe.As`

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+
+            for (nint index = 0; index < Count; index++)
+            {
+                T value = ScalarGreaterThanOrEqual(left.GetElement(index), right.GetElement(index)) ? GetAllBitsSetValue() : default;
```
This was previously doing `Equals(left, right) | GreaterThan(left, right)`, but I changed it to an explicit `ScalarGreaterThanOrEqual` helper. Same goes for `LessThanOrEqual`

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+            ThrowHelper.ThrowForUnsupportedVectorBaseType<T>();
+            Vector<T> result = default;
+
+            result._00 = (left._00 & condition._00) | (right._00 & ~condition._00);
```
This was doing `(left & condition) | (Vector.AndNot(right, condition))`, which themselves just did a direct op on the underlying `long` field.
I've simplified it to directly doing the relevant ops on the two long fields.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector_1.cs:
```diff
@@ -0,0 +1,1507 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool ScalarEquals(T left, T right)
+        {
+            if (typeof(T) == typeof(byte))
```
This is where the actual per-type logic lives, same as previously for most of the methods.

tannergooding committed:
Fixing System.Numerics.Vectors.Tests

tannergooding committed:
Adding back Register so morph doesn't break

tannergooding commented:
CC. @CarolEidt, @pgovind, @jkotas

jkotas commented on src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:
```diff
@@ -466,6 +466,9 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Numerics\Matrix4x4.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Numerics\Plane.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Numerics\Quaternion.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Numerics\Register.cs" />
```
Why do we still need Register.cs ?

tannergooding replied:
> I had to add back Register. ARM64 was failing due to some disconnect between the HFA classification causing it to be `hasMultiregStructArgs` and `fgMorphMultiregStructArgs` not reporting the type as TYP_STRUCT

I don't want to block the overall simplification on this and am going to do some deeper investigation as to why. @CarolEidt might also have some insight as I believe she added much of the HFA support.

jkotas replied:
ok, I agree it is fine to look at this separately.

CarolEidt replied:
It's not immediately apparent to me what the issue would be with the HFA classification, but agree that it would be best to address that separately.

CarolEidt reviewed (commented):
It's a bit difficult to review given the name changes; might be easier to review if the filename changes were done as a separate commit? I'll complete my review on Monday.

tannergooding committed:
Removing the T4 templates used by Vector<T>

tannergooding committed:
Renaming Vector.cs to Vector_1.cs

tannergooding committed:
Renaming Vector_Operations.cs to Vector.cs

tannergooding commented:
Rebased the changes so the renames happen separately from the code changes, as per @CarolEidt's comment: https://github.com/dotnet/runtime/pull/42050#pullrequestreview-487174403

Dropped the commit that was adjusting the `operator *(Vector<T>, T)` case as per @jkotas' comment: https://github.com/dotnet/runtime/pull/42050#discussion_r487328586

CarolEidt reviewed (approved):
LGTM - thanks so much for separating out the renames; makes it much easier to review the actual changes!
